src/object_store/src/object/mod.rs

```diff
@@ -933,5 +999,178 @@ pub async fn build_remote_object_store(
 pub type ObjectDataStream = BoxStream<'static, ObjectResult<Bytes>>;
+
+#[derive(Debug, Clone, Copy)]
+enum OperationType {
+    Upload,
+    StreamingUploadInit,
+    StreamingUpload,
+    StreamingUploadFinish,
+    Read,
+    StreamingReadInit,
+    StreamingRead,
+    Metadata,
+    Delete,
+    DeleteObjects,
+    List,
+}
+
+impl OperationType {
+    fn as_str(&self) -> &'static str {
+        match self {
+            Self::Upload => "upload",
+            Self::StreamingUploadInit => "streaming_upload_init",
+            Self::StreamingUpload => "streaming_upload",
+            Self::StreamingUploadFinish => "streaming_upload_finish",
+            Self::Read => "read",
+            Self::StreamingReadInit => "streaming_read_init",
+            Self::StreamingRead => "streaming_read",
+            Self::Metadata => "metadata",
+            Self::Delete => "delete",
+            Self::DeleteObjects => "delete_objects",
+            Self::List => "list",
+        }
+    }
+}
+
+fn get_retry_attempts_by_type(config: &ObjectStoreConfig, operation_type: OperationType) -> usize {
+    match operation_type {
+        OperationType::Upload => config.retry.upload_retry_attempts,
+        OperationType::StreamingUploadInit
+        | OperationType::StreamingUpload
+        | OperationType::StreamingUploadFinish => config.retry.streaming_upload_retry_attempts,
+        OperationType::Read => config.retry.read_retry_attempts,
+        OperationType::StreamingReadInit | OperationType::StreamingRead => {
+            config.retry.streaming_read_retry_attempts
+        }
+        OperationType::Metadata => config.retry.metadata_retry_attempts,
+        OperationType::Delete => config.retry.delete_retry_attempts,
+        OperationType::DeleteObjects => config.retry.delete_objects_retry_attempts,
+        OperationType::List => config.retry.list_retry_attempts,
+    }
+}
+
+fn get_attempt_timeout_by_type(config: &ObjectStoreConfig, operation_type: OperationType) -> u64 {
+    match operation_type {
+        OperationType::Upload => config.retry.upload_attempt_timeout_ms,
+        OperationType::StreamingUploadInit | OperationType::StreamingUpload => {
+            config.retry.streaming_upload_attempt_timeout_ms
+        }
+        OperationType::StreamingUploadFinish => {
+            config.retry.streaming_upload_finish_attempt_timeout_ms
+        }
+        OperationType::Read => config.retry.read_attempt_timeout_ms,
+        OperationType::StreamingReadInit | OperationType::StreamingRead => {
+            config.retry.streaming_read_attempt_timeout_ms
+        }
+        OperationType::Metadata => config.retry.metadata_attempt_timeout_ms,
+        OperationType::Delete => config.retry.delete_attempt_timeout_ms,
+        OperationType::DeleteObjects => config.retry.delete_attempt_timeout_ms,
+        OperationType::List => config.retry.list_attempt_timeout_ms,
+    }
+}
+
+struct RetryCondition {
+    max_duration_ms: u64,
```

Comment:
https://github.com/risingwavelabs/risingwave/pull/16231#discussion_r1570656358. We don't need to maintain `max_duration_ms` if we don't want to ensure backward compat.